pkg/tls/tls_test.go

```diff
@@ -0,0 +1,241 @@
+	serialNumber, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 128))
+	if err != nil {
+		fmt.Errorf("failed to create cert serial number")
+	}
+	return serialNumber, nil
+}
+func getTemplate() (*x509.Certificate, error) {
+	now := time.Now()
+	SN, err := getSerialNumber()
+	if err != nil {
+		fmt.Errorf("Failed to create cert serial number")
+	}
+
+	template := &x509.Certificate{
+		SerialNumber:          SN,
+		NotBefore:             now,
+		NotAfter:              now.AddDate(5, 0, 0),
+		Subject:               pkix.Name{CommonName: cn, Organization: orgs},
+		Issuer:                pkix.Name{CommonName: cn, Organization: orgs},
+		BasicConstraintsValid: true,
+		// the field below is ignored
+		//KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		IsCA: true,
+	}
+	return template, nil
+}
+
+//this is supposed to be rarely used
+func createKnownGoodCert(isCA bool) *x509.Certificate {
+	template := &x509.Certificate{
+		SerialNumber:          big.NewInt(0),
+		NotBefore:             time.Unix(0, 0),
+		NotAfter:              time.Unix(0, 0),
+		Subject:               pkix.Name{CommonName: cn, Organization: orgs},
+		Issuer:                pkix.Name{CommonName: cn},
+		IPAddresses:           []net.IP{net.ParseIP("192.168.0.1")},
+		DNSNames:              []string{cn},
+		BasicConstraintsValid: true,
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth | x509.ExtKeyUsageClientAuth},
+		IsCA:                  isCA,
+	}
+	// let's make a key real quick
+	pk, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		fmt.Errorf("Error generating temp private key %s", err)
+	}
+	knownGoodDER, err := x509.CreateCertificate(rand.Reader, template, template, pk.Public(), pk)
+	if err != nil {
+		fmt.Errorf("Error generating known good cert %s", err)
+	}
+	knownGood, err := x509.ParseCertificate(knownGoodDER)
+	if err != nil {
+		fmt.Errorf("unable to parse DER %s", err)
+	}
+	return knownGood
+}
+
+func loadGoodCert(path string) *x509.Certificate {
+
+	knownGoodBlob, err := ioutil.ReadFile(path)
+	if err != nil {
+		fmt.Errorf("unable to load last known good cert from disk %s", err)
+	}
+	block, _ := pem.Decode(knownGoodBlob)
+	if block == nil || block.Type != "CERTIFICATE" {
+		fmt.Errorf("failed to decode PEM block containing public key")
+	}
+
+	knownGoodCerts, err := x509.ParseCertificates(block.Bytes)
+	if err != nil {
+		fmt.Errorf("unable to parse certificates %s", err)
+	}
+	goodCert := knownGoodCerts[0]
+	return goodCert
+}
+
+const cn = "dummy-test-certificate.local"
+const knownGoodLoc = "./testdata/known_good_cert.pem"
+const knownGoodLocCA = "./testdata/known_good_certCA.pem"
+const updateSSLVar = "UPDATE_KNOWN_SSL_CERT"
+
+var orgs = []string{cn}
+
+func TestNewPrivateKey(t *testing.T) {
+	newPK, err := NewPrivateKey()
+	if err != nil {
+		t.Errorf("failed to create private key")
```

Comment:
no need to go to the effort of `t.Errorf("some new error message")` in tests.  The line number is already reported for you if the test fails.  Use `t.Error(err)` or `t.Fatal(err)` as a general rule.

Reply:
understood